bin/reflective_loader/src/loader.rs

```diff
@@ -0,0 +1,845 @@
+        let mut relocation_entry_ptr: *mut u16 = (relocation_block_ref as usize + core::mem::size_of::<IMAGE_BASE_RELOCATION>() as usize) as *mut u16;
+        for _index in 0..relocation_block_entries_count {
+            let relocation_entry: BaseRelocationEntry = BaseRelocationEntry::new(unsafe{*relocation_entry_ptr});
+            if relocation_entry.reloc_type as u32 == IMAGE_REL_BASED_DIR64 || relocation_entry.reloc_type as u32 == IMAGE_REL_BASED_HIGHLOW {
+                let addr_to_be_patched = (new_dll_base as usize + relocation_block.VirtualAddress as usize + relocation_entry.offset as usize) as *mut usize;
+                let new_value_at_addr  = unsafe { *addr_to_be_patched } + image_base_delta as usize;
+                unsafe { *addr_to_be_patched = new_value_at_addr };
+            }
+            relocation_entry_ptr = (relocation_entry_ptr as usize + BaseRelocationEntry::c_size()) as *mut u16;
+        }
+        relocation_block_ref = (relocation_block_ref as usize + relocation_block.SizeOfBlock as usize) as *mut IMAGE_BASE_RELOCATION;
+    }
+}
+
+/// AND the ILT entry (a 64 or 32 bit value) by the b10000000... to get the most signifacnt bit.
+/// Check if that most significant bit is 0 or 1. 
+/// If it's 1 then the function should be loaded by ordinal reference.   - return True
+/// If it's 0 then the function should be loaded by name.                - return False
+fn image_snap_by_ordinal(ordinal: usize) -> bool{
+    #[cfg(target_arch = "x86_64")]
+    return (ordinal as u64 & IMAGE_ORDINAL_FLAG64) != 0;
+}
+
+/// Extract the 0-15 bytes which represent the ordinal
+/// reference to import the function with.
+/// C variation: `def IMAGE_ORDINAL(Ordinal): return (Ordinal & 0xffff)`
+fn image_ordinal(ordinal: usize) -> u16 {
+    return (ordinal & 0xffff) as u16;
+}
+
+fn process_import_address_tables(new_dll_base: *mut c_void, pe_file_headers: &PeFileHeaders64, load_library_a_fn: FnLoadLibraryA, get_proc_address_fn: FnGetProcAddress, get_last_error_fn: FnGetLastError) -> () {
+    let import_directory: IMAGE_DATA_DIRECTORY = pe_file_headers.nt_headers.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT as usize];
+	
+    if import_directory.Size == 0 {
+        // No relocations to process
+        return;
+    }
+
+    let mut base_image_import_table: *mut IMAGE_IMPORT_DESCRIPTOR = (new_dll_base as usize + import_directory.VirtualAddress as usize) as *mut IMAGE_IMPORT_DESCRIPTOR;
+    loop {
+        let import_table_descriptor = unsafe{*base_image_import_table};
+        if import_table_descriptor.Name == 0 {
+            break;
+        }
+
+        let library_name_ptr = (new_dll_base as usize + import_table_descriptor.Name as usize) as *const i8;
+        let library_name = unsafe { CStr::from_ptr(library_name_ptr) };
+        let library_handle = load_library_a(load_library_a_fn, get_last_error_fn, library_name.as_ptr() as *const u8);
+        if library_handle != 0 {
```

Comment:
Should I just panic if load library fails?

Reply:
is the panic in a separate process from golem/imix?

Reply:
Found it! The panic occurs in the loader so the shim that realm hands execution off to.
We don’t have a good way to get output or errors from this any failures will be a black box unless we implement some sort of IPC comms.